### ISSUE TITLE
openssh.listenAddresses cause an infinite recursion

### ISSUE THREAD

**Issue by robberer** [NixOS/nixpkgs#7354]
Hi all,

with current master branch my nixos modules do not work anymore. I get an "infinite recursion encountered"

`git bisect` tells me that commit 1d62ad474694c0717017c2c8aa79909a890407b5 is the evil one.

My modules look like this:

http://nixpaste.noip.me/raw/Ulg5mEaj

And here is the `--show-trace` output:

http://nixpaste.noip.me/raw/Oa5NzJIt

I import the module in my configuration.nix like this:

```
imports =
    [
      ./hardware-configuration.nix
      ./snmpd.nix
      ./kled-hdd.nix
      ./kled-net.nix
      ./btrsnap.nix
      ./dovecot.nix
      ./fetch-nixos-irc.nix
      ./htpasswd-file.nix
      ./containers.nix
      ./mosix.nix
    ];
```

@shlevy

**Comment by robberer** [comment]
OK, after reading PR #6794 I changed `{config, pkgs, ...}:` to `{config, pkgs, lib, ...}:` and `with pkgs.lib;` to `with lib;`

Now I get a new error: `error: anonymous function at "/etc/nixos/nixpkgs/nixos/modules/services/networking/ssh/sshd.nix":382:29 called with unexpected argument ‘_module’, at undefined position`

Is this another issue or a new follow-up error.

**Comment by robberer** [comment]
This is how I have sshd set up in configuration.nix:

```
# Enable the OpenSSH daemon.
    openssh = {
      enable = true;
      listenAddresses = [ { addr = "192.168.77.222"; port = 22; } { addr = "1.1.1.1"; port = 64022; } ];
    };
```

**Comment by shlevy** [comment]
@nbp

**Comment by copumpkin** [comment]
I'm also getting infinite recursion on a config running against then nixos-unstable branch (on the channel branches). The trace looks different, so perhaps it's a different bug. I'll look into it some more.

**Comment by copumpkin** [comment]
Yeah, pretty sure anything running config against nixos-unstable just unconditionally gets infinite recursion. I thought there were tests?

**Comment by copumpkin** [comment]
Adding the blocker tag because I think it basically breaks all of nixos-unstable. Tell me if you have it working and we can try something else.

**Comment by nbp** [comment]
@robberer, This is sadly an expected behaviour which is that `_module.check` (or its old alias) should not be defined with any other option definitions, but only with an unconditional value.

**Comment by nbp** [comment]
@copumpkin @robberer  can you link to a minimal configuration which reproduce this issue?

**Comment by robberer** [comment]
@nbp here is a minimal config that reproduce the sshd error. no custom modules included. I can reproduce this on multiple real hosts.

http://nixpaste.noip.me/raw/rxWrMt9u

**Comment by robberer** [comment]
But I guess, just copy and paste the openssh service part to _any_ config is enough to reproduce

**Comment by robberer** [comment]
@copumpkin Do you get "infinite recursion encountered" with self written modules outside of nixpkgs tree ? If so, did you applied the changes mentioned in PR #6794 ?

With those hints - especially the "mkOption, mkIf" part, I'm not getting "infinite recursion encountered" with other "custom" modules so far.

@nbp 

But I noticed at least one more error related to #6794.

Hint: Apache SSL Virtual Hosts .

So I guess the mentioned PR is bigger regression than I thought. It seems to break almost any common config out there.

**Comment by nbp** [comment]
> So I guess the mentioned PR is bigger regression than I thought. It seems to break almost any common config out there.

I don't think so, at least hydra VM tests in addition to many other configuration files are working fine.
I will fix it as soon as I find some time to do it. (hopefully this week)

Sorry for the delay.

**Comment by robberer** [comment]
Don't worry. I'm glad to hear that someone is about to investigate this issue. :+1:

**Comment by MP2E** [comment]
For what it's worth, I have a fairly non-trivial config with many derivations defined outside of nixpkgs and everything is still working fine here.

However, I don't have any SSH support defined in my config.

**Comment by nbp** [comment]
Here is the minimal configuration with which the module system fails:

```
{
  boot.loader.grub.device = "/dev/sda";
  fileSystems."/".label = "root";
  services.openssh = {
    enable = true;
    listenAddresses = [ { addr = "192.168.77.222"; port = 22; } { addr = "1.1.1.1"; port = 64022; } ];
  };
}
```

Removing any line in this configuration change the error message, or make the configuration work.

**Comment by nbp** [comment]
Ok, I think I found the source of the issue.  I will make a patch tomorrow as I have to catch a train.
To make it short, in `sshd.nix` we map the result of  `listenAddresses` with a function which explictly expects 2 arguments, but no `_module` argument.

The solution is to strip the `_module` attribute, or to use ellipsis.

**Comment by copumpkin** [comment]
My failure didn't explicitly touch sshd but I did have a an extraUsers
entry with an ssh key. Would that also infinitely recurse?

**Comment by nbp** [comment]
@robberer, I pushed the following branch in the mean time.

https://github.com/nbp/nixpkgs-2/tree/issue-7354

@copumpkin , I will look at your issue with extraUsers, but this is likely similar.

**Comment by copumpkin** [comment]
@nbp thanks! if it ends up being hard to repro I'll try to trim down my configuration to a useful case, but it sounds like we're on the right path already

**Comment by cillianderoiste** [comment]
I also got an "infinite recursion encountered" error after trying to upgrade to nixos-15.05pre60707.4e51a46 but it was caused by the following entry in my hardware-configuration.nix:

    require = [
      "${modulesPath}/installer/scan/not-detected.nix"
    ];

This was fixed by changing it to:

    imports = [
      <nixpkgs/nixos/modules/installer/scan/not-detected.nix>
    ];

I have no idea if this is related, but perhaps someone else runs into this too.

**Comment by nbp** [comment]
@copumpkin , I checked other use cases, I was not able to find a similar pattern where we use an attribute set deconstruction on a `subsmodule`/`optionSet`.  Would you be able to make a minimal test case out of your configuration?

**Comment by nbp** [comment]
@cillianderoiste , hum this is unfortunate, knowing that we were generating such files.  This fix was suggested as part of #6794.  I will add this to the release notes, but then we should probably revert the modulePath change at least to display a message when it is used.

**Comment by nbp** [comment]
@copumpkin , I will close this issue, feel free to open a new one for extraUsers.
